extmod/modframebuf.c

```diff
@@ -97,13 +97,69 @@ STATIC void rgb565_fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, i
     }
 }
 
+// Functions for GRAY4 format
+
+STATIC void gray4_setpixel(const mp_obj_framebuf_t *fb, int x, int y, uint32_t col) {
+    uint8_t *pixel = &((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1];
+
+    if (x % 2) {
+        *pixel = ((uint8_t)col & 0x0F) | (*pixel & 0xF0);
+    }
+    else {
+        *pixel = ((uint8_t)col << 4) | (0x0F & *pixel);
+    }
+}
+
+STATIC uint32_t gray4_getpixel(const mp_obj_framebuf_t *fb, int x, int y) {
+    if (x % 2) {
+        return ((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1] & 0x0F;
+    }
+
+    return ((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1] >> 4;
+}
+
+STATIC void gray4_fill_rect(const mp_obj_framebuf_t *fb, int x, int y, int w, int h, uint32_t col) {
+    uint8_t *pixel_pair = &((uint8_t*)fb->buf)[(x + y * fb->stride) >> 1];
+    uint8_t col_cleaned = (uint8_t)col & 0x0F;
+    uint8_t col_shifted_left = col_cleaned << 4;
+    uint8_t colored_pixel_pair = col_shifted_left | col_cleaned;
```

Comment:
I don't think you gain much by having these explicit variables.  Can you check if the code size reduces (or stays the same) if you just repeat these in the code below?  You can just do `col &= 0x0f` here and then use `col << 4` below.